java/ql/src/experimental/Security/CWE/CWE-297/IgnoredHostnameVerification.ql

```diff
@@ -0,0 +1,40 @@
+/**
+ * @name Ignored result of hostname verification
+ * @description The method HostnameVerifier.verify() returns a result of hostname verification.
+ *              A caller has to check the result and drop the connection if the verification failed.
+ * @kind problem
+ * @problem.severity error
+ * @precision high
+ * @id java/ignored-hostname-verification
+ * @tags security
+ *       external/cwe/cwe-297
+ */
+
+import java
+
+/** The `HostnameVerifier.verify()` method. */
+private class HostnameVerifierVerifyMethod extends Method {
+  HostnameVerifierVerifyMethod() {
+    this.getDeclaringType().getASupertype*().hasQualifiedName("javax.net.ssl", "HostnameVerifier") and
+    this.hasStringSignature("verify(String, SSLSession)")
+  }
+}
+
+/** Defines `HostnameVerifier.verity()` calls that is not wrapped in another `HostnameVerifier`. */
+private class HostnameVerificationCall extends MethodAccess {
+  HostnameVerificationCall() {
+    this.getMethod() instanceof HostnameVerifierVerifyMethod and
+    not this.getCaller() instanceof HostnameVerifierVerifyMethod
+  }
+
+  /** Holds if the result of the call is not used. */
+  predicate isIgnored() {
+    not exists(Expr expr, IfStmt ifStmt, MethodAccess ma |
```

Comment:
This is relying on the CodeQL optimiser not to compute the product of all Exprs, IfStmts and MethodAccesses because there are disjuncts where each of them are in scope but unused.

Instead, do something like

```
not exists(Expr e | this = e.getAChildExpr()) and 
not exists(IfStmt ifStmt | ...
```

However you can also determine if a method call is unused like this by checking whether it's the direct child of an ExprStmt--

```
  this = any(ExprStmt es).getExpr()
```

All other kinds of use (assignment, passing to a method call, direct use in a conditional...) will be nested within some other statement or expression.

Reply:
Thanks for the suggestions!